### PR TITLE
docs(deeplinker): add inital docs

### PR DESCRIPTION
**Pull request by mhartington**
#### Short description of what this resolves:

Still a few changes that I need to make regarding dynamic links
- [x] Simple Link segment
- [x] Dynamic params
- [x] Default history

**Ionic Version**: 1.x / 2.x
2.x

### CODE REVIEW

**Comment by mhartington** [comment]
Didn't get to change the location strategy, but that can be added down the road.

File: src/navigation/deep-linker.ts [hunk 1]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ * - History: Deep linker's string array of internal URL history
+ * - Location: Angular's Location provider, which abstracts Hash/Path Location Strategies
+ *
+ * @usage
+ *
+ * DeepLinker can be used in the `IonicModule.forRoot` method, as the third parameter
+ *
+ * ```ts
+ *  imports: [
+ *     IonicModule.forRoot(MyApp, {} {
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: Missing a comma after the 2nd parameter

File: src/navigation/deep-linker.ts [hunk 2]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ *       links: []
+ *    })
+ *   ]
+ * ```
+ *
+ * DeepLinker implements `DeepLinkerConfig`, is an object with an links array. So for basic example based on the blank starer, a link setup like so:
+ *
+ * ```ts
+ *  imports: [
+ *     IonicModule.forRoot(MyApp, {} {
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: Missing comma here, as well

File: src/navigation/deep-linker.ts [hunk 3]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ *       links: []
+ *    })
+ *   ]
+ * ```
+ *
+ * DeepLinker implements `DeepLinkerConfig`, is an object with an links array. So for basic example based on the blank starer, a link setup like so:
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: `an links array` -> `a links array`

File: src/navigation/deep-linker.ts [hunk 4]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ *  imports: [
+ *     IonicModule.forRoot(MyApp, {} {
+ *       links: [
+ *        { component: HomePage, name: 'Home', segment: 'home' }
+ *      ]
+ *    })
+ *   ]
+ * ```
+ *
+ * This Feels pretty familiar to how Angular sets up routes, but how some fundamental differences.
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: `but how some fundamental differences` should `how` be `have`?

File: src/navigation/deep-linker.ts [hunk 5]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ * So at any point, when a Component becomes the active view, we just append the URL segment.
+ *
+ * ### Dynamic Links
+ *
+ * Since passing data around is common practice in an app, we can reflect that in our app's URL in a similar manner to Angular's router.
+ *
+ * ```ts
+ *  links: [
+ *    { component: HomePage, name: 'Home', segment: 'home' }
+ *    { component: DetailPage, name: 'Detail', segment: 'detal/:user' }
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: typo should be `detail/:user` ?

File: src/navigation/deep-linker.ts [hunk 6]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ *  ]
+ *  ```
+ * This approach of using `:param` has been around in previous routing solutions.
+ * All this means is that when we push a new component on to the stack, in the navParams, there should be a property of of `user`.
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: double `of of`

File: src/navigation/deep-linker.ts [hunk 7]
```diff
@@ -45,35 +28,124 @@ import { ViewController } from './view-controller';
...
+ *  In some cases when a page loads, you might be sent to a component that has it's own information, but not back view.
+ *  This situation is common when loading a page from a Push Notification.
+ *  If you want a component to have a default history when none is present, you can use the `defaultHistory` property
+ *
+ * The `defaultHistory` property takes an array of components to create the history stack if none exist.
+ *
+ * ```ts
+ *  links: [
+ *    { component: HomePage, name: 'Home', segment: 'home' }
+ *    { component: DetailPage, name: 'Detail', segment: 'detal/:user', defaultHistory: [HomePage] }
```
Comment (brandyscarney) [inline src/navigation/deep-linker.ts]: `detal/:user` -> `detail/:user`

File: src/navigation/deep-linker.ts (hunk 1 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: fixed

File: src/navigation/deep-linker.ts (hunk 3 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: fixed

File: src/navigation/deep-linker.ts (hunk 2 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: fixed

File: src/navigation/deep-linker.ts (hunk 4 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: fixed

File: src/navigation/deep-linker.ts (hunk 7 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: fixed

File: src/navigation/deep-linker.ts [hunk 8]
```diff
@@ -45,35 +28,84 @@ import { ViewController } from './view-controller';
  * Ionic developers to think of URLs as a breadcrumb rather than as the source of
  * truth in navigation. This encourages flexible navigation design and happy apps all
  * over the world.
- */
-
-/**
- * @private
  *
+ * Deep Linking Scenarios:
+ * 1. Initialize all NavControllers using the initial browser URL
```
Comment (mhartington) [inline src/navigation/deep-linker.ts]: This was in there originally, didn't know if it should be removed. Will remove though

File: src/navigation/deep-linker.ts [hunk 9]
```diff
@@ -86,6 +72,8 @@ import { ViewController } from './view-controller';
  *  ```
  * This approach of using `:param` has been around in previous routing solutions.
  * All this means is that when we push a new component on to the stack, in the navParams, there should be a property of `user`.
+ * The property needs to be something that can be serialize by the DeepLinker.
```
Comment (mlynch) [inline src/navigation/deep-linker.ts]: forgot a d on serialized

File: src/navigation/deep-linker.ts [hunk 10]
```diff
@@ -70,7 +56,7 @@ import { ViewController } from './view-controller';
  * ```
  *
  * This Feels pretty familiar to how Angular sets up routes, but has some fundamental differences.
```
Comment (mlynch) [inline src/navigation/deep-linker.ts]: There's not enough information in the rest of the doc for this sentence to be proven so let's just take it out

Comment (mhartington) [inline src/navigation/deep-linker.ts]: Fixed

File: src/navigation/deep-linker.ts (hunk 9 above)
Comment (mhartington) [inline src/navigation/deep-linker.ts]: Fixed

**Pull request merged**